File: app/src/main/java/org/example/StreamableProtoFileWriter.java
package org.example;

import java.io.BufferedOutputStream;
import java.io.DataOutputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.function.BiConsumer;

public class StreamableProtoFileWriter<H extends com.google.protobuf.GeneratedMessageV3, P extends com.google.protobuf.GeneratedMessageV3>
        implements AutoCloseable {

    private final DataOutputStream fi;
    private boolean sealed = false;
    private BiConsumer<Long, P> onWriteCallback;
    private long currentWriteOffset = 0;
    
    public  StreamableProtoFileWriter(String file, H header, BiConsumer<Long, P> onWriteCallback)
            throws FileNotFoundException, IOException, IllegalArgumentException {      
                this(file, header);
                this.onWriteCallback = onWriteCallback;
    }
    public StreamableProtoFileWriter(String file, H header)
            throws FileNotFoundException, IOException, IllegalArgumentException {

        if (file == null) {
            throw new IllegalArgumentException("File path cannot be null");
        }

        if (header == null) {
            throw new IllegalArgumentException("Header cannot be null");
        }
        if (!file.endsWith(".binpb")) {
            throw new IllegalArgumentException("File must end with .binpb");
        }
        this.fi = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(file), 64 * 1024));

        this.fi.writeInt(StreamableProtoFileParser.MAGIC_BYTE);
        this.currentWriteOffset += 4;
        
        byte[] headerBytes = header.toByteArray();
        
        this.fi.writeInt(headerBytes.length);
        this.currentWriteOffset += 4;
        this.fi.write(headerBytes);
        this.currentWriteOffset += headerBytes.length;

    }

    public int Write(P payload) throws IOException {

        if (sealed) {
            throw new IOException("File is sealed");
        }

        byte[] payloadBytes = payload.toByteArray();

        if (this.onWriteCallback != null) {
            this.onWriteCallback.accept(this.currentWriteOffset, payload);
        }
        this.fi.writeInt(payloadBytes.length);
        this.currentWriteOffset += 4;
        this.fi.write(payloadBytes);
        this.currentWriteOffset += payloadBytes.length;

        return payloadBytes.length;
    }
    // Seal the file
    public void Seal() throws IOException {
        if (!sealed) {
            this.fi.writeInt(StreamableProtoFileParser.FILE_SEAL_MARKER);
            sealed = true;
        }
    }

    public void close() throws Exception {

        Seal();

        fi.flush();
        fi.close();
    }
}

File: python-app/proto_streamed_file/StreamableProtoFileWriter.py
import struct
from . import StreamableProtoFileParser
from google.protobuf.message import Message


class StreamableProtoFileWriter:
    
    def __init__(self,file_name, header_proto_instance):
        if not isinstance(header_proto_instance, Message):
            raise ValueError("header_proto_instance must be a protobuf message")
        
        if file_name is None:
            raise ValueError("file_name cannot be None")
        if not file_name.endswith('.binpb'):
            raise ValueError("file_name must end with .binpb")
        
        self.file_name = file_name
        self.file = open(self.file_name, "wb")
      
        magic_byte = struct.pack('>i',StreamableProtoFileParser.StreamableProtoFileParser.MAGIC_BYTE)
        self.file.write(magic_byte)  
        header_bytes = header_proto_instance.SerializeToString()
        header_size = struct.pack('>i', len(header_bytes))
        self.file.write(header_size)
        self.file.write(header_bytes)
        self.sealed = False

    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.seal()
        self.file.close()
    
    
    def writePayload(self, payload_proto_instance):
        if self.sealed:
            raise ValueError("Cannot write payload after seal")
        
        if not isinstance(payload_proto_instance, Message):
            raise ValueError("payload_proto_instance must be a protobuf message")  
        payload_bytes = payload_proto_instance.SerializeToString()
        payload_size = struct.pack('>i', len(payload_bytes))
        self.file.write(payload_size)
        self.file.write(payload_bytes)
        return payload_size
    
    def seal(self):
        if self.sealed:
            return
        seal = struct.pack('>i', StreamableProtoFileParser.StreamableProtoFileParser.FILE_SEAL_MARKER)
        self.file.write(seal)
        self.sealed = True